persist/seeds_proposer_test.go

```diff
@@ -0,0 +1,114 @@
+package persist
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/test"
+	kb "github.com/libp2p/go-libp2p-kbucket"
+	"github.com/libp2p/go-libp2p-peerstore"
+	swarmt "github.com/libp2p/go-libp2p-swarm/testing"
+	bhost "github.com/libp2p/go-libp2p/p2p/host/basic"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestRandomSeedsProposer(t *testing.T) {
+	testCases := map[string]struct {
+		nTotalCandidates           int // snapshotted candidate list
+		nNonCandidatePeersInRT     int // non candidate peers that are already in the routing table
+		nCandidatesNotInPeerStore  int // candidates that are absent in the peer store
+		nCandidatesAlreadyInRT     int // candidates that are "not-diallable"/not alive
+		nFallbacks                 int // fallback list
+		nFallbacksAlreadyInRT      int // fallback peers that are already in RT
+		expectedNumPeersInProposal int // number of peers we expect in the routing table after seeding is complete
+	}{
+		"No Proposals -> RT already has target peers": {10, 10, 0, 0, 10, 0, 0},
+
+		"No Proposals -> candidate & fallback sets are empty": {0, 0, 0, 0, 0, 0, 0},
+
+		"No Proposals -> candidates & fallbacks already in RT": {3, 0, 0, 3, 3, 3, 0},
+
+		"No Proposals -> candidates not in peerstore & fallbacks already in RT": {3, 0, 3, 0, 3, 3, 0},
+
+		"Success -> only candidates": {15, 2, 5, 2, 0, 0, 6},
+
+		"Success -> candidates + fallbacks": {5, 2, 1, 1, 1, 0, 4},
+
+		"Success -> only fallbacks": {7, 2, 5, 2, 10, 2, 4},
+	}
+
+	for name, testcase := range testCases {
+		// create host for self
+		self := bhost.New(swarmt.GenSwarm(t, context.Background(), swarmt.OptDisableReuseport))
+		target := 10
+
+		// create candidate hosts & add them to the peer store
+		candidateHosts := make([]bhost.BasicHost, testcase.nTotalCandidates)
+		candidatePids := make([]peer.ID, testcase.nTotalCandidates)
+		for i := 0; i < testcase.nTotalCandidates; i++ {
+			h := *bhost.New(swarmt.GenSwarm(t, context.Background(), swarmt.OptDisableReuseport))
```

Comment:
use ctx once you have the cancellable context

Reply:
Done